test/webgpu_basic_rendering.cpp

```diff
@@ -16,7 +18,34 @@
 
 #include <emscripten.h>
 #include <emscripten/html5.h>
-#include <emscripten/html5_webgpu.h>
+#include <emscripten/em_js.h>
+
+EM_JS_DEPS(deps, "$keepRuntimeAlive");
+
+// Keeps track of whether async tests are still alive to make sure they finish
+// before exit. This tests that keepalives exist where they should.
+static int sScopeCount = 0;
+class ScopedCounter {
+    public:
+        ScopedCounter(const ScopedCounter&&) { Increment(); }
+        ScopedCounter(const ScopedCounter&) { Increment(); }
+        ScopedCounter() { Increment(); }
+        ~ScopedCounter() { Decrement(); }
+    private:
+        void Increment() { sScopeCount++; }
+        void Decrement() { assert(sScopeCount > 0); sScopeCount--; }
+};
+
+void RegisterCheckScopesAtExit() {
+    atexit([](){
+        // Check we don't exit before the tests are done.
+        // (Make sure there's a keepalive for everything the test has scopes for.)
+        // Build with -sEXIT_RUNTIME to trace keepalives.
+        assert(sScopeCount == 0);
+        // Overwrite the old return code and exit now that everything is done.
+        exit(0);
```

Comment:
I think it's undefined behavior to call `exit()` from inside an `atexit()` handler. (ISO/IEC 9899:n2347.pdf, §7.22.4.4/2: "If a program calls the exit function more than once, the behavior is undefined.")